### PR TITLE
Bluetooth scanner version 2

### PR DESCRIPTION
**Pull request by taulfsime**
In this script version, a queue is established to store packets, and a timer is employed to flush the queue.
Additionally, it addresses the issue where the device owns the scan process, but the script attempts to initiate a new one, which causes the script to not work.

### CODE REVIEW

**Comment by taulfsime** [comment]
Issue: https://github.com/home-assistant-libs/aioshelly/issues/477

**Comment by taulfsime** [comment]
Issue: https://github.com/home-assistant-libs/aioshelly/issues/477

File: aioshelly/ble/const.py [hunk 1]
```diff
@@ -10,31 +10,63 @@
...
+  
+  for(let i = 0; i < burstSendCount; i++) {
+    if (queue.length <= 0) {
+      break;
     }
-});
-BLE.Scanner.Start({
-    duration_ms: %duration_ms%,
+
+    Shelly.emitEvent("%event_type%", queue[0]);
```
Comment (bdraco) [inline aioshelly/ble/const.py:31]: Can we change the event to include multiple payloads at once?  That should reduce the amount of small packets the device has to send

Comment (bdraco) [reply aioshelly/ble/const.py:31]: In ESPHome we do `Bluetooth advertisement bundling` which means we try to send 16 (32 if the device has SPIRAM at a time to minimize the time the wifi radio is on.

If you can do something similar here, I can take care of refactoring the integration code to do something similar to https://github.com/Bluetooth-Devices/bleak-esphome/blob/b93b9eec29b1183e72ffb5a0510891b328076743/src/bleak_esphome/backend/scanner.py#L40

Comment (bdraco) [reply aioshelly/ble/const.py:31]: That also means we could update https://www.home-assistant.io/integrations/bluetooth/#remote-adapters-bluetooth-proxies to list Bluetooth advertisement bundling support for shelly

File: aioshelly/ble/const.py [hunk 2]
```diff
@@ -10,31 +10,63 @@
...
+  }
+}
+
+// Skip starting if scanner is active
+if (!BLE.Scanner.isRunning()) {
+  BLE.Scanner.Start({
+    duration_ms: -1,
     active: %active%,
-    interval_ms: %interval_ms%,
-    window_ms: %window_ms%,
```
Comment (bdraco) [inline aioshelly/ble/const.py:38]: Why drop the interval and window?

**Review by bdraco — CHANGES_REQUESTED** [review]
Thanks @taulfsime 

I think this is a good change.  Please see comment above

File: aioshelly/ble/const.py [hunk 3]
```diff
@@ -10,31 +10,63 @@
 VAR_EVENT_TYPE = "%event_type%"
 VAR_ACTIVE = "%active%"
 VAR_VERSION = "%version%"
-VAR_INTERVAL_MS = "%interval_ms%"
-VAR_WINDOW_MS = "%window_ms%"
-VAR_DURATION_MS = "%duration_ms%"
```
Comment (chemelli74) [inline aioshelly/ble/const.py:15]: Those are currently used in `aioshelly/ble/__init__.py`

File: aioshelly/ble/const.py [hunk 4]
```diff
@@ -10,31 +10,63 @@
...
-
-DEFAULT_INTERVAL_MS = 320
-DEFAULT_WINDOW_MS = 30
-DEFAULT_DURATION_MS = -1
```
Comment (chemelli74) [inline aioshelly/ble/const.py:19]: Those are currently used by HA integration: `shelly/bluetooth/__init__.py`

**Comment by chemelli74** [comment]
Applied this PR to my production instance that suffer script memory overflow for Shelly BLU Gateway (`hw: SNGW-BT01`, `fw  20231219-133956/1.1.0-g34b5d4f`)

File: aioshelly/ble/const.py (hunk 2 above)
Comment (taulfsime) [reply aioshelly/ble/const.py]: I thought the default could be used, but it's better to keep it as it was. I just bring them back.

File: aioshelly/ble/const.py (hunk 4 above)
Comment (taulfsime) [reply aioshelly/ble/const.py]: Yeah, sorry, I thought the default config could be used, but yeah. I just bring them back.

File: aioshelly/ble/const.py (hunk 1 above)
Comment (taulfsime) [reply aioshelly/ble/const.py]: With https://github.com/home-assistant-libs/aioshelly/pull/478/commits/619ce095c29ebf8c94372267f22e96fda5830699 the script will send data as an object now with 2 props, first one is "ver" - response version and the second one "advs" which is a list of arrays with the same data as before. 

The parser method `parse_ble_scan_result_event` now returns list with objects, and each object contains "address", "rssi" and "adv" keys.

Comment (taulfsime) [reply aioshelly/ble/const.py]: @bdraco Should I open a MR in the core repo to update the handler?

File: aioshelly/ble/const.py [hunk 5]
```diff
@@ -19,22 +19,68 @@
...
+      break;
     }
-});
-BLE.Scanner.Start({
+
+    Shelly.emitEvent(
+      "%event_type%", {
+        ver: %version%, 
+        advs: queue.slice(0, packetsInSingleEvent),
+      }
```
Comment (bdraco) [inline aioshelly/ble/const.py:43]:
```suggestion
      "%event_type%", [
        %version%, 
        queue.slice(0, packetsInSingleEvent),
      ]
```

**Comment by bdraco** [comment]
I did some tweaks to make it backwards compatible

**Comment by bdraco** [comment]
I've just about got it wired up to `habluetooth` so it can be used in a dev env outside of HA.  I'll take care of the core PR shortly

**Comment by bdraco** [comment]
https://github.com/home-assistant/core/pull/107917

File: aioshelly/ble/const.py (hunk 1 above)
Comment (bdraco) [reply aioshelly/ble/const.py]: I took care of it in https://github.com/home-assistant/core/pull/107917

**Review by bdraco — APPROVED** [review]

File: aioshelly/ble/const.py [hunk 6]
```diff
@@ -19,22 +19,68 @@
...
-            res.addr,
-            res.rssi,
-            btoa(res.advData),
-            btoa(res.scanRsp)
-        ]);
+// aioshelly BLE script 2.0
+const queueServeTimer = 450; // in ms, timer for events emitting
+const burstSendCount =  5; // number if events, emitted on timer event
+const maxQueue =  80; // if the queue exceeds the limit, all new events are ignored until it empties
+const packetsInSingleEvent = 5; // max number of packets in single event
```
Comment (bdraco) [inline aioshelly/ble/const.py:26]: @taulfsime Can we put 16 (to match ESPHome) in an event or is that too large?

Comment (bdraco) [reply aioshelly/ble/const.py:26]:
```
2024-01-12 13:47:57.890 WARNING (MainThread) [aioshelly.ble.backend.scanner] parsed_advs: [('55:F2:94:5F:18:42', -47, (None, [], {}, {76: b'\t\x08\x13\xf1\xc0\xa8k\xbd\x1bX\x16\x08\x00\xda-\xed\xdd%h\\'}, None)), ('D4:B2:A1:7C:42:17', -64, ('RZSS', [], {}, {}, None)), ('C5:F0:B6:05:9A:B8', -65, (None, [], {}, {76: b'\x12\x02$\x01'}, None)), ('F6:EE:7E:AC:9F:EB', -55, (None, [], {}, {76: b'\x12\x02.\x03\x07\x11\x06(\x89\x8amE\x88C!\xde/\xa1#WD\xae\xa9'}, None)), ('FC:EE:36:CF:93:95', -68, (None, [], {}, {2409: b'\xfc\xee6\xcf\x93\x95\x13\x1fd\x12\x04'}, None)), ('D8:EF:2F:41:B1:34', -72, ('RZSS', [], {}, {}, None)), ('34:AB:95:85:66:6D', -74, ('LOOKin_98F33163', ['00001812-0000-1000-8000-00805f9b34fb'], {}, {}, None))]
2024-01-12 13:47:58.453 WARNING (MainThread) [aioshelly.ble.backend.scanner] parsed_advs: [('74:CD:9C:23:FC:C4', -43, (None, [], {}, {76: b'\x10\x05!\x98_oC'}, 8)), ('24:0B:2A:36:55:4B', -72, (None, [], {}, {76: b'\t\x08\x13\x08\xc0\xa8k[\x1bX'}, None)), ('63:8F:51:DD:CC:0C', -58, (None, [], {}, {76: b'\x13\x0b\x1b|\x02\xa4\xc4_?6\xa1\xbe\x00'}, None)), ('E2:88:93:E1:50:6E', -44, (None, [], {}, {76: b'\x12\x02\x00\x00'}, None)), ('67:AE:22:73:84:52', -55, (None, [], {}, {76: b'\x10\x070\x1f\x08\xcf\n^h'}, 12)), ('F8:04:2E:E1:9F:19', -74, (None, [], {}, {117: b'B\x04\x01\x80\xa0\xf8\x04.\xe1\x9f\x19\xfa\x04.\xe1\x9f\x18\x01\x00\x00\x00\x00\x00\x00'}, None))]
2024-01-12 13:47:59.516 WARNING (MainThread) [aioshelly.ble.backend.scanner] parsed_advs: [('D1:4B:33:1F:0A:FD', -69, (None, [], {}, {2409: b'\xd1K3\x1f\n\xfd?\x1dd\x11\x04'}, None)), ('68:60:7C:2F:C1:1B', -76, (None, [], {}, {76: b'\t\x08\x13\xc5\xc0\xa8k~\x1bX\x16\x08\x00\xe8\xad\x12\xca\xfeIW'}, None)), ('F1:D0:17:BA:C6:A2', -75, ('Nanoleaf Strip 3B32', [], {}, {}, None)), ('34:AB:95:85:6E:F9', -68, ('LOOKin_98F330B4', ['00001812-0000-1000-8000-00805f9b34fb'], {}, {}, None)), ('5A:DB:5D:53:1E:84', -48, (None, [], {}, {76: b'\x0c\x0e\x08\xd9\\,\xe0\x91\x94\x9a\xf2^\xf5F\xd8\xf8\x10\x06A\x1dR\xdf\xe9\x18'}, None)), ('6A:1D:DC:AD:36:E2', -47, (None, [], {}, {76: b'\x0c\x0e\x08\xd9\\,\xe0\x91\x94\x9a\xf2^\xf5F\xd8\xf8\x10\x06A\x1dR\xdf\xe9\x18'}, None)), ('CC:32:37:35:4E:05', -59, ('GVH5106_4E05', ['0000ec88-0000-1000-8000-00805f9b34fb'], {}, {1: b'\x01\x01\x0e\xa3s\xa8'}, None))]
2024-01-12 13:48:00.364 WARNING (MainThread) [aioshelly.ble.backend.scanner] parsed_advs: [('C0:95:6D:53:56:AD', -71, (None, [], {}, {76: b'\x10\x05\n\x14\x84\xfa\xce'}, 12)), ('7A:56:85:6A:6D:CD', -60, (None, ['0000fe07-0000-1000-8000-00805f9b34fb'], {}, {1447: b'\x05\x10\x01\x00\x00\x00\x00\x00\x00\x02#\x00\xca'}, 9)), ('6A:1D:DC:AD:36:E2', -42, (None, [], {}, {76: b'\x0c\x0e\x08\xda\\h\xab\xfd\x9f\xd7YcF_\xe3\x08\x10\x06A\x1dR\xdf\xe9\x18'}, None)), ('5A:DB:5D:53:1E:84', -42, (None, [], {}, {76: b'\x0c\x0e\x08\xda\\h\xab\xfd\x9f\xd7YcF_\xe3\x08\x10\x06A\x1dR\xdf\xe9\x18'}, None))]
2024-01-12 13:48:01.107 WARNING (MainThread) [aioshelly.ble.backend.scanner] parsed_advs: [('D4:B2:A1:7C:42:17', -71, ('RZSS', [], {}, {}, None)), ('FC:EE:36:CF:93:95', -71, (None, [], {}, {2409: b'\xfc\xee6\xcf\x93\x95\x13\x1fd\x12\x04'}, None)), ('34:AB:95:85:66:6D', -71, ('LOOKin_98F33163', ['00001812-0000-1000-8000-00805f9b34fb'], {}, {}, None)), ('55:F2:94:5F:18:42', -37, (None, [], {}, {76: b'\t\x08\x13\xf1\xc0\xa8k\xbd\x1bX\x16\x08\x00\xda-\xed\xdd%h\\'}, None))]
True
2024-01-12 13:48:01.670 WARNING (MainThread) [aioshelly.ble.backend.scanner] parsed_advs: [('CC:32:37:35:4E:05', -59, ('GVH5106_4E05', ['0000ec88-0000-1000-8000-00805f9b34fb'], {}, {1: b'\x01\x01\x0e\x94-\x80'}, None)), ('D8:EF:2F:41:B1:34', -70, ('RZSS', [], {}, {}, None)), ('74:CD:9C:23:FC:C4', -52, (None, [], {}, {76: b'\x10\x05!\x98_oC'}, 8)), ('63:8F:51:DD:CC:0C', -45, (None, [], {}, {76: b'\x13\x0b\x1b|\x02\xa4\xc4_?6\xa1\xbe\x00'}, None))]
```

It _seems_ to work ok to go to 16

Comment (bdraco) [inline aioshelly/ble/const.py:26]:
```suggestion
const packetsInSingleEvent = 16; // max number of packets in single event
```

Comment (bdraco) [reply aioshelly/ble/const.py:26]: I practice I never seem to get more than 9

File: setup.py [hunk 7]
```diff
@@ -5,7 +5,7 @@
 
 PROJECT_DIR = Path(__file__).parent.resolve()
 README_FILE = PROJECT_DIR / "README.md"
-VERSION = "7.1.0"
+VERSION = "8.0.0"
```
Comment (bdraco) [inline setup.py:8]: Needs to be reverted before merging

**Comment by bdraco** [comment]
diagnostics look good overnight

File: aioshelly/ble/const.py [hunk 8]
```diff
@@ -19,22 +19,68 @@
...
-    if (ev === BLE.Scanner.SCAN_RESULT) {
-        Shelly.emitEvent("%event_type%", [
-            %version%,
-            res.addr,
-            res.rssi,
-            btoa(res.advData),
-            btoa(res.scanRsp)
-        ]);
+// aioshelly BLE script 2.0
+const queueServeTimer = 450; // in ms, timer for events emitting
```
Comment (thecode) [inline aioshelly/ble/const.py:23]: Can we shorten this interval or try to think of a logic which will not require a big delay?
One of the advantages I liked with the Shelly BLE + Proxy is that the bluetooth response was very fast, with this change there is a noticeable delay with the events.
I use Shelly with door sensor to switch lights and buttons to switch monitors on a PC and the delay is noticeable

Comment (bdraco) [inline aioshelly/ble/const.py:23]:
```suggestion
const queueServeTimer = 100; // in ms, timer for events emitting
```

**Comment by bdraco** [comment]
Retesting with the above timing tweaks.  We can revert them if needed

**Review by thecode — APPROVED** [review]

**Review by bieniu — APPROVED** [review]
Tested for several hours and works without problems, thanks 👍

**Comment by bdraco** [comment]
Thanks @taulfsime 

We will get this shipped in 2024.2.x

**Pull request merged**